### PR TITLE
Image url validation

### LINKED ISSUE #4: Image URL Validation
#### As a user I want to be prevented from submitting invalid image URLs.

Acceptance criteria:
- [ ] I cannot successfully save an image with an invalid URL.
- [ ] An error message is associated with the appropriate input field on
  failure.

Discussion Topic:
- [ ] What is a valid image URL?

Dependencies:
- Save Image Link

### PR DESCRIPTION
**Pull request by tlconnor**
Closes #4

### CODE REVIEW

File: app/validators/image_url_validator.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
+#
+# Validates an image URL.  The format of the URL is checked and an optional HTTP HEAD request can be made
+# to check that the HTTP response is of an image mime type.
+#
+class ImageUrlValidator < ActiveModel::EachValidator
+  def validate_each(record, attribute, value)
+    valid = check_url_format(value)
+    valid = valid && check_exists(value) if options.key?(:check_exists)
```
Comment (bboe) [inline app/validators/image_url_validator.rb]: Just in case you weren't aware, ruby has a `&&=` operator.

File: app/validators/image_url_validator.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+      uri = URI.parse(url)
+      uri.kind_of?(URI::HTTP)
+    rescue URI::InvalidURIError
+    end
+  end
+
+  def check_exists(url)
+    begin
+      res = Faraday.head(url)
+    rescue Faraday::ConnectionFailed
```
Comment (bboe) [inline app/validators/image_url_validator.rb]: Are there other exceptions that might need to be handled?

Also, as we discussed yesterday, I think if we use this method, there should be a timeout to avoid stalling the request more than some acceptable amount of time.

File: app/validators/image_url_validator.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    begin
+      uri = URI.parse(url)
+      uri.kind_of?(URI::HTTP)
+    rescue URI::InvalidURIError
+    end
+  end
+
+  def check_exists(url)
+    begin
+      res = Faraday.head(url)
```
Comment (bboe) [inline app/validators/image_url_validator.rb]: Can we use a more descriptive variable than than `res`?

File: config/application.rb [hunk 4]
```diff
@@ -51,5 +51,8 @@ class Application < Rails::Application
     # Middleware like session, flash, cookies can be added back manually.
     # Skip views, helpers and assets when generating a new resource.
     config.api_only = true
+
+    # Add custom validators path
+    config.autoload_paths << Rails.root.join('app', 'validators')
```
Comment (bboe) [inline config/application.rb]: Awesome.

File: test/validators/image_url_validator_test.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
+require 'test_helper'
+
+class MockModel
```
Comment (bboe) [inline test/validators/image_url_validator_test.rb]: Awesome.

File: test/validators/image_url_validator_test.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+class MockModelWithExistanceCheck
+  include ActiveModel::Validations
+  attr_accessor :url
+  validates :url, image_url: { check_exists: true }
+end
+
+class ImageUrlValidatorTest < ActiveSupport::TestCase
+  def test_valid_url_without_existance_check
+    model = MockModel.new
+    model.url = 'http://url.that.does.not.exist'
```
Comment (bboe) [inline test/validators/image_url_validator_test.rb]: While this should fail pretty quickly as the `exist` top-level-domain does not exist, should we mock out the networking component?

File: test/validators/image_url_validator_test.rb [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    assert_predicate model, :valid?
+  end
+
+  def test_valid_url_with_existance_check
+    model = MockModelWithExistanceCheck.new
```
Comment (bboe) [inline test/validators/image_url_validator_test.rb]: While this should fail pretty quickly as the `exist` top-level-domain does not exist, should we mock out the networking component anyway in the event that it does exist in the future?

File: test/validators/image_url_validator_test.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    model.url = 'http://url.that.does.not.exist'
+    assert_predicate model, :invalid?
+    assert_equal ['is not a valid image URL'], model.errors['url']
```
Comment (bboe) [inline test/validators/image_url_validator_test.rb]: This assertion should definitely be mocked as we cannot guarantee that the image will always be there, or the server always be up.

Also, I think it makes the tests more readable if they do just one thing. Can we split this method into a success and failure case?

File: test/validators/image_url_validator_test.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    model.url = 'https://www.appfolio.com/images/html/apm-logo-v2.png'
+    assert_predicate model, :valid?
+  end
+
+  def test_blank_url
```
Comment (bboe) [inline test/validators/image_url_validator_test.rb]: Should we allow blank/nil in this validator, and rely on the user to add the `presence` validator if they want to require that?

If not, then I think we should also have a test for nil here.

File: test/validators/image_url_validator_test.rb [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+  end
+
+  def test_non_http_url
+    model = MockModel.new
+    model.url = 'ftp://something'
+    assert_predicate model, :invalid?
+    assert_equal ['is not a valid image URL'], model.errors['url']
+  end
+
+  def test_url_without_protocol
```
Comment (bboe) [inline test/validators/image_url_validator_test.rb]: When not testing the URL, what about urls like:

    https:/www.appfolio.com/images/html/apm-logo-v2.png

File: app/models/image.rb [hunk 11]
```diff
@@ -1,3 +1,3 @@
 class Image < ApplicationRecord
-  validates :url, presence: true
+  validates :url, image_url: true
```
Comment (bboe) [inline app/models/image.rb]: If this isn't going to use the HEAD request component, why add it?

File: app/components/api.js [hunk 12]
```diff
@@ -1,12 +1,26 @@
...
  * For 200 responses a Promise for the JSON is returned.  Otherwise an error is thrown
  */
 function checkResponseStatus(res) {
-  if (res.status !== 200) {
-    throw Error(res.statusText);
+  if (res.status === 422) {
+    return res.json().then(data => {
+      throw new ValidationError(data.errors);
+    });
+  } else if (res.status < 200 || res.status >= 300) {
```
Comment (bboe) [inline app/components/api.js]: I wonder what you'll be using those other 2XX responses for. :)

File: test/components/api.spec.js [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+    it('returns a promise for the returned image data', () => (
+      api.getImages().then(data => expect(data).to.deep.equal(images))
+    ));
+  });
+
+  describe('#getImage', () => {
+    let successResponse;
+    let notFoundResponse;
+    beforeEach(() => {
```
Comment (bboe) [inline test/components/api.spec.js]: Do we need to do both of these before each test?

File: test/components/api.spec.js [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+    ));
+
+    it('returns a promise for the returned image data', () => (
+      api.getImage(1).then(data => expect(data).to.deep.equal(images[0]))
+    ));
+
+    it('returns a rejected promise if the image is not found', () => (
+      api
+        .getImage(2)
+        .then(() => {
```
Comment (bboe) [inline test/components/api.spec.js]: ⭐  Great test.

File: test/components/api.spec.js [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+        .reply(422, JSON.stringify({
+          status: 422, error: 'Unprocessable Entity', errors: { url: ['is not a valid image URL'] }
+        }));
+
+      successResponse = nock('https://example.appfolio.com')
+        .post('/images', { url: 'http://example.url' })
+        .reply(201, JSON.stringify({ image: images[2] }));
+    });
+
+    it('makes a request to the API', () => (
```
Comment (bboe) [inline test/components/api.spec.js]: Would the other tests fail if `isDone` is false? In other words, are these tests essential?

File: app/components/form/field.js [hunk 16]
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
```
Comment (bboe) [inline app/components/form/field.js]: These are awesome.

File: app/components/form/field.js [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+  render() {
+    const { children, name, value, error, onChange } = this.props;
+
+    const glyphClass = classnames('glyphicon', 'form-control-feedback', error && 'glyphicon-remove');
+
+    return (
+      <div className={classnames('form-group', error && 'has-error has-feedback')}>
+        <label htmlFor={name} className="col-sm-2 control-label">{children}</label>
+        <div className="col-sm-6">
+          <input id={name} name={name} className="form-control" value={value} type="text" onChange={onChange} />
```
Comment (bboe) [inline app/components/form/field.js]: Should the `id` be prefixed at all in the event that name is not unique to other IDs that are rendered on the page? Or is this concern irrelevant due to the automatic name spacing?

File: test/components/form/button.spec.js [hunk 18]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import { shallow } from 'enzyme';
+import { expect } from 'chai';
+import sinon from 'sinon';
+
+import Button from 'components/form/button';
+
+describe('<Button />', () => {
```
Comment (bboe) [inline test/components/form/button.spec.js]: ⭐  great tests

File: app/components/images/show.js [hunk 19]
```diff
@@ -29,7 +30,7 @@ export default class ShowImage extends React.Component {
       <div className={styles.image}>
         Image {imageId}
         <img src={image.url} />
-        <a className="btn btn-primary" href="/">Back</a>
+        <Link className="btn btn-secondary" to="/">Back</Link>
```
Comment (bboe) [inline app/components/images/show.js]: Will this always be a back button? Or is it always a "home" button?

Also what do you get by changing it from an explicit "a" to a react Link?

File: app/components/images/new.js [hunk 20]
```diff
@@ -14,31 +14,39 @@ export class NewImage extends React.Component {
...
 
   submit(event) {
+    this.setState({ loading: true });
     event.preventDefault();
 
     createImage({ url: this.state.url })
       .then(data => {
         this.props.router.push(`/images/${data.id}`);
+      })
+      .catch(e => {
```
Comment (bboe) [inline app/components/images/new.js]: Is there a way to catch a certain class of exceptions? I imagine not all exceptions have an `errors` attribute.

**Comment by bboe** [comment]
Added my initial set of feedback and questions. Looks like there are a few lint issues that circleci is reporting on, and the heroku app doesn't seem to be happy yet.

For the most part things look great, and I really am liking the react components.

File: app/validators/image_url_validator.rb (hunk 1 above)
Comment (tlconnor) [inline app/validators/image_url_validator.rb]: Ah yeah, its been a while since I've worked with ruby!

File: app/validators/image_url_validator.rb (hunk 2 above)
Comment (tlconnor) [inline app/validators/image_url_validator.rb]: I've added a timeout and changed the rescue block to catch the more generic `Faraday::ClientError` that covers pretty much any exception relating to HTTP / network

File: app/validators/image_url_validator.rb (hunk 3 above)
Comment (tlconnor) [inline app/validators/image_url_validator.rb]: fixed

File: test/validators/image_url_validator_test.rb (hunk 7 above)
Comment (tlconnor) [inline test/validators/image_url_validator_test.rb]: I've added the `webmock` gem for HTTP mocking.  It now mocks all requests by default.  In the tests I've now added specific mocking for the various success/failure cases

File: test/validators/image_url_validator_test.rb (hunk 8 above)
Comment (tlconnor) [inline test/validators/image_url_validator_test.rb]: as per my previous comment, now fixed

File: test/validators/image_url_validator_test.rb (hunk 9 above)
Comment (tlconnor) [inline test/validators/image_url_validator_test.rb]: Good point.  I've added a separate `validates :url, presence: true` and now skip blank during the custom validator

File: test/validators/image_url_validator_test.rb (hunk 10 above)
Comment (tlconnor) [inline test/validators/image_url_validator_test.rb]: added

File: app/models/image.rb (hunk 11 above)
Comment (tlconnor) [inline app/models/image.rb]: good point, now fixed

File: test/components/api.spec.js (hunk 13 above)
Comment (tlconnor) [inline test/components/api.spec.js]: Probably not, refactored to be cleaner

File: test/components/api.spec.js (hunk 15 above)
Comment (tlconnor) [inline test/components/api.spec.js]: I've not quite sure what you mean here.

File: app/components/form/field.js (hunk 17 above)
Comment (tlconnor) [inline app/components/form/field.js]: The `id` probably should be prefixed at some stage but I didn't want to add the extra complexity until it was needed.

File: app/components/images/show.js (hunk 19 above)
Comment (tlconnor) [inline app/components/images/show.js]: Not sure, depends on future requirements.  I added it simply as a way to get back.

The reason I switched from `<a>` to `<Link>` is because Link takes advantage of the router and the history API.  An `<a>` will take the user to `/` (with a page refresh) while a `<Link>` will take the user to `/#/` without any refresh

File: app/components/images/new.js (hunk 20 above)
Comment (tlconnor) [inline app/components/images/new.js]: There is no way to specify exception types in javascript.  The best we can do is something like
```
if(e typeof ValidationError) {
 ...
}
```
in the catch block.  In this case we don't really care; if `.errors` is present it is used.

File: test/components/api.spec.js (hunk 15 above)
Comment (bboe) [inline test/components/api.spec.js]: These `makes a request to the API` tests seem to be an implicit part of the test that actually extracts the data. Thus it seems they might be redundant. I'm asking, are they redundant?

File: app/components/images/show.js (hunk 19 above)
Comment (bboe) [inline app/components/images/show.js]: Oh that's right (about not refreshing). Love that.

> I added it simply as a way to get back.

Could we name it 'Home' to better reflect that one is always going to `/`? If the show view is linked from somewhere else this "back" link will be misleading.

File: app/components/images/new.js (hunk 20 above)
Comment (bboe) [inline app/components/images/new.js]: I wasn't aware of that. Shows how much javascript I write. This looks fine then.

File: app/models/image.rb [hunk 21]
```diff
@@ -1,3 +1,9 @@
 class Image < ApplicationRecord
-  validates :url, image_url: true
+  validates :url, presence: true
+
+  # On create check that the image URL exists
+  validates :url, image_url: { check_exists: true }, on: :create
+
+  # On update check that the image URL format is valid
+  validates :url, image_url: true, on: :update
```
Comment (bboe) [inline app/models/image.rb]: For the purpose of atomic commits I think we should save this addition until update is added.

File: app/validators/image_url_validator.rb [hunk 22]
```diff
@@ -4,27 +4,28 @@
...
   def validate_each(record, attribute, value)
-    valid = check_url_format(value)
-    valid = valid && check_exists(value) if options.key?(:check_exists)
-    record.errors[attribute] << 'is not a valid image URL' unless valid
+    return if value.blank?
+
+    if !check_url_format(value)
+      record.errors[attribute] << 'is not a valid image URL'
+    elsif options.key?(:check_exists) && !check_exists(value)
+      record.errors[attribute] << 'does not exist'
```
Comment (bboe) [inline app/validators/image_url_validator.rb]: Does this error message, and the function name for that matter, convey what is actually being checked for? While checking for a 404 is one of the possibilities, it's also possible that the content-type isn't an image.

File: app/validators/image_url_validator.rb [hunk 23]
```diff
@@ -4,27 +4,28 @@
...
   def check_exists(url)
-    begin
-      res = Faraday.head(url)
-    rescue Faraday::ConnectionFailed
-    end
-
-    res && res.headers['content-type'] =~ /^image/
+    http_response = Faraday::Connection.new.head(url) { |request| request.options.timeout = 2 }
+    http_response.headers['Content-Type'] =~ /^image/
+  rescue Faraday::ClientError
```
Comment (bboe) [inline app/validators/image_url_validator.rb]: 👍

File: test/controllers/images_controller_test.rb [hunk 24]
```diff
@@ -1,6 +1,14 @@
 require 'test_helper'
 
 class ImagesControllerTest < ActionDispatch::IntegrationTest
+  setup do
```
Comment (bboe) [inline test/controllers/images_controller_test.rb]: Similar question. Do we need these stubs in all our test functions?

File: test/models/image_test.rb [hunk 25]
```diff
@@ -2,20 +2,22 @@
 
 class ImageTest < ActiveSupport::TestCase
   def test_valid_image
-    image = valid_image
+    stub_request(:head, 'http://image.url')
```
Comment (bboe) [inline test/models/image_test.rb]: I see how this validator is annoying now since it always needs to be stubbed out.

File: test/validators/image_url_validator_test.rb [hunk 26]
```diff
@@ -13,38 +13,45 @@ class MockModelWithExistanceCheck
...
-    assert_predicate model, :valid?
   end
 
-  def test_blank_url
+  def test_url_without_protocol
     model = MockModel.new
-    assert_predicate model, :invalid?
-    assert_equal ['is not a valid image URL'], model.errors['url']
+
+    %w('//something', 'http:/something', 'http:something').each do |url|
```
Comment (bboe) [inline test/validators/image_url_validator_test.rb]: I'm curious how this test is passing. Maybe it was fixed in a later version:

```
2.2.3 :002 > require 'uri'
 => true 
2.2.3 :003 > URI.parse('http:/something')
 => #<URI::HTTP http:/something> 
```

**Comment by bboe** [comment]
Second round of feedback complete. I like the timeout handling a lot, and the mocking -- though that does seem cumbersome to have to deal with anytime we need to create an image model with validations.

File: app/models/image.rb (hunk 21 above)
Comment (tlconnor) [inline app/models/image.rb]: I would argue that this *is* atomic.  The atomic action is adding `url` validation.

File: app/validators/image_url_validator.rb (hunk 22 above)
Comment (tlconnor) [inline app/validators/image_url_validator.rb]: Its validating that the URL points to an image.  I don't see the need to distinguish between 404 and non-image.

File: test/controllers/images_controller_test.rb (hunk 24 above)
Comment (tlconnor) [inline test/controllers/images_controller_test.rb]: I figured they were fair enough.   There are already multiple tests calling `Image.create!` and will likely be more in the future.  This way we don't need to repeat the same stubs for every test.

File: test/models/image_test.rb (hunk 25 above)
Comment (tlconnor) [inline test/models/image_test.rb]: Yes, I'm tempted to forget about the network test thing :(

File: app/models/image.rb (hunk 21 above)
Comment (bboe) [inline app/models/image.rb]: That's a good point. It does make validation more complete, though, this adds code that for something that doesn't yet exist, and as a result it's not being tested so that breaks the test all the things rule.

File: app/components/images/show.js (hunk 19 above)
Comment (tlconnor) [inline app/components/images/show.js]: I could do that, but I've already deleted it in the next PR I'm working on (style app)

Comment (bboe) [inline app/components/images/show.js]: Fair enough.

**Comment by tlconnor** [comment]
@bboe I have folded the code review changes back into the original commits.  Could you please take a final look so we can get this merged?

File: app/models/image.rb [hunk 27]
```diff
@@ -1,3 +1,4 @@
 class Image < ApplicationRecord
   validates :url, presence: true
+  validates :url, image_url: true
```
Comment (bboe) [inline app/models/image.rb]: These two statements, since there are no conditionals, can be specified in one line.

File: test/validators/image_url_validator_test.rb (hunk 10 above)
Comment (bboe) [inline test/validators/image_url_validator_test.rb]: It looks like these were removed wit the refactor. Can you ensure those tests are there?

File: test/components/api.spec.js (hunk 15 above)
Comment (bboe) [inline test/components/api.spec.js]: Any update to this question?

> Would the other tests fail if isDone is false? In other words, are these tests essential?

> These makes a request to the API tests seem to be an implicit part of the tests that actually extracts the data (e.g., returns a promise for the returned image data). Thus it seems they might be redundant. I'm asking, are they redundant?

File: app/models/image.rb (hunk 27 above)
Comment (tlconnor) [inline app/models/image.rb]: Fixed
